File: internal/services/task_service.go
package services

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/cloudstreet-dev/REST-gRPC-in-33-Languages/code/go/internal/models"
	"github.com/google/uuid"
)

// TaskService provides task management operations
type TaskService struct {
	mu    sync.RWMutex
	tasks map[string]*models.Task
}

// NewTaskService creates a new task service instance
func NewTaskService() *TaskService {
	s := &TaskService{
		tasks: make(map[string]*models.Task),
	}
	s.initializeSampleData()
	return s
}

// initializeSampleData creates some sample tasks
func (s *TaskService) initializeSampleData() {
	sampleTasks := []models.CreateTaskRequest{
		{
			Title:       "Implement Go REST API",
			Description: "Create REST API with Gin framework",
			Priority:    models.TaskPriorityHigh,
			Tags:        []string{"go", "rest", "api"},
		},
		{
			Title:       "Add gRPC support",
			Description: "Implement gRPC server with native Go support",
			Priority:    models.TaskPriorityMedium,
			Tags:        []string{"go", "grpc", "protobuf"},
		},
	}

	for _, req := range sampleTasks {
		s.CreateTask(req)
	}
}

// ListTasks returns a paginated list of tasks with optional filters
func (s *TaskService) ListTasks(query models.ListTasksQuery) (*models.ListTasksResponse, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	// Convert map to slice for filtering and sorting
	var tasks []*models.Task
	for _, task := range s.tasks {
		tasks = append(tasks, task)
	}

	// Apply filters
	tasks = s.filterTasks(tasks, query)

	// Apply sorting
	s.sortTasks(tasks, query.SortOrder)

	// Apply pagination
	totalCount := len(tasks)
	pageSize := query.PageSize
	if pageSize <= 0 {
		pageSize = 20
	}
	if pageSize > 100 {
		pageSize = 100
	}

	startIndex := 0
	if query.PageToken != "" {
		if idx, err := strconv.Atoi(query.PageToken); err == nil {
			startIndex = idx
		}
	}

	endIndex := startIndex + pageSize
	if endIndex > totalCount {
		endIndex = totalCount
	}

	var paginatedTasks []models.Task
	if startIndex < totalCount {
		for _, task := range tasks[startIndex:endIndex] {
			paginatedTasks = append(paginatedTasks, *task)
		}
	}

	var nextPageToken string
	if endIndex < totalCount {
		nextPageToken = strconv.Itoa(endIndex)
	}

	return &models.ListTasksResponse{
		Tasks:         paginatedTasks,
		NextPageToken: nextPageToken,
		TotalCount:    totalCount,
	}, nil
}

// GetTask retrieves a task by ID
func (s *TaskService) GetTask(id string) (*models.Task, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	task, exists := s.tasks[id]
	if !exists {
		return nil, fmt.Errorf("task with ID %s not found", id)
	}

	return task, nil
}

// CreateTask creates a new task
func (s *TaskService) CreateTask(req models.CreateTaskRequest) (*models.Task, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	now := time.Now()
	task := &models.Task{
		ID:          uuid.New().String(),
		Title:       req.Title,
		Description: req.Description,
		Status:      models.TaskStatusPending,
		Priority:    req.Priority,
		Tags:        req.Tags,
		CreatedBy:   "system",
		AssignedTo:  req.AssignedTo,
		CreatedAt:   now,
		UpdatedAt:   now,
		DueDate:     req.DueDate,
	}

	// Set default priority if not specified
	if task.Priority == "" {
		task.Priority = models.TaskPriorityMedium
	}

	// Validate priority
	if !task.Priority.IsValid() {
		return nil, fmt.Errorf("invalid priority: %s", task.Priority)
	}

	// Initialize empty tags slice if nil
	if task.Tags == nil {
		task.Tags = []string{}
	}

	s.tasks[task.ID] = task
	return task, nil
}

// UpdateTask updates an existing task
func (s *TaskService) UpdateTask(id string, req models.UpdateTaskRequest) (*models.Task, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	task, exists := s.tasks[id]
	if !exists {
		return nil, fmt.Errorf("task with ID %s not found", id)
	}

	// Update fields if provided
	if req.Title != nil {
		task.Title = *req.Title
	}
	if req.Description != nil {
		task.Description = *req.Description
	}
	if req.Status != nil {
		if !req.Status.IsValid() {
			return nil, fmt.Errorf("invalid status: %s", *req.Status)
		}
		task.Status = *req.Status
		
		// Set completed time if status changes to completed
		if task.Status == models.TaskStatusCompleted && task.CompletedAt == nil {
			now := time.Now()
			task.CompletedAt = &now
		}
	}
	if req.Priority != nil {
		if !req.Priority.IsValid() {
			return nil, fmt.Errorf("invalid priority: %s", *req.Priority)
		}
		task.Priority = *req.Priority
	}
	if req.Tags != nil {
		task.Tags = req.Tags
	}
	if req.AssignedTo != nil {
		task.AssignedTo = req.AssignedTo
	}
	if req.DueDate != nil {
		task.DueDate = req.DueDate
	}

	task.UpdatedAt = time.Now()
	return task, nil
}

// UpdateTaskStatus updates the status of a task
func (s *TaskService) UpdateTaskStatus(id string, status models.TaskStatus) (*models.Task, error) {
	return s.UpdateTask(id, models.UpdateTaskRequest{
		Status: &status,
	})
}

// DeleteTask deletes a task by ID
func (s *TaskService) DeleteTask(id string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, exists := s.tasks[id]; !exists {
		return fmt.Errorf("task with ID %s not found", id)
	}

	delete(s.tasks, id)
	return nil
}

// filterTasks applies filters to the task list
func (s *TaskService) filterTasks(tasks []*models.Task, query models.ListTasksQuery) []*models.Task {
	var filtered []*models.Task

	for _, task := range tasks {
		// Filter by status
		if query.Status != "" && task.Status != query.Status {
			continue
		}

		// Filter by assigned user
		if query.AssignedTo != "" {
			if task.AssignedTo == nil || *task.AssignedTo != query.AssignedTo {
				continue
			}
		}

		// Filter by tags
		if query.Tags != "" {
			tags := strings.Split(query.Tags, ",")
			hasAllTags := true
			for _, tag := range tags {
				tag = strings.TrimSpace(tag)
				found := false
				for _, taskTag := range task.Tags {
					if taskTag == tag {
						found = true
						break
					}
				}
				if !found {
					hasAllTags = false
					break
				}
			}
			if !hasAllTags {
				continue
			}
		}

		filtered = append(filtered, task)
	}

	return filtered
}

// sortTasks sorts tasks based on the specified order
func (s *TaskService) sortTasks(tasks []*models.Task, order models.SortOrder) {
	switch order {
	case models.SortOrderCreatedAtAsc:
		sort.Slice(tasks, func(i, j int) bool {
			return tasks[i].CreatedAt.Before(tasks[j].CreatedAt)
		})
	case models.SortOrderCreatedAtDesc:
		sort.Slice(tasks, func(i, j int) bool {
			return tasks[i].CreatedAt.After(tasks[j].CreatedAt)
		})
	case models.SortOrderDueDateAsc:
		sort.Slice(tasks, func(i, j int) bool {
			if tasks[i].DueDate == nil {
				return false
			}
			if tasks[j].DueDate == nil {
				return true
			}
			return tasks[i].DueDate.Before(*tasks[j].DueDate)
		})
	case models.SortOrderDueDateDesc:
		sort.Slice(tasks, func(i, j int) bool {
			if tasks[i].DueDate == nil {
				return false
			}
			if tasks[j].DueDate == nil {
				return true
			}
			return tasks[i].DueDate.After(*tasks[j].DueDate)
		})
	case models.SortOrderPriorityAsc:
		sort.Slice(tasks, func(i, j int) bool {
			return tasks[i].Priority.GetValue() < tasks[j].Priority.GetValue()
		})
	case models.SortOrderPriorityDesc:
		sort.Slice(tasks, func(i, j int) bool {
			return tasks[i].Priority.GetValue() > tasks[j].Priority.GetValue()
		})
	}
}

File: src/TaskManagementApi/Services/TaskService.cs
using TaskManagementApi.Models;
using TaskManagementApi.Models.Requests;
using TaskManagementApi.Models.Responses;

namespace TaskManagementApi.Services;

/// <summary>
/// Provides task management operations with in-memory storage.
/// Uses ReaderWriterLockSlim to match Go's sync.RWMutex semantics:
/// concurrent reads are allowed, writes are exclusive.
/// </summary>
public class TaskService
{
    private readonly ReaderWriterLockSlim _lock = new();
    private readonly Dictionary<string, TaskModel> _tasks = new();

    public TaskService()
    {
        InitializeSampleData();
    }

    /// <summary>
    /// Creates a new task with defaults and validation.
    /// Mirrors Go's CreateTask: generates UUID, sets status=pending,
    /// defaults priority to medium if not specified, ensures tags is non-null.
    /// </summary>
    public TaskModel CreateTask(CreateTaskRequest req)
    {
        _lock.EnterWriteLock();
        try
        {
            var now = DateTime.UtcNow;
            var priority = req.Priority ?? TaskPriority.Medium;

            // Validate priority
            if (!priority.IsValid())
            {
                throw new ValidationException($"invalid priority: {priority.ToSnakeCase()}");
            }

            var task = new TaskModel
            {
                Id = Guid.NewGuid().ToString(),
                Title = req.Title,
                Description = req.Description,
                Status = Models.TaskStatus.Pending,
                Priority = priority,
                Tags = req.Tags ?? new List<string>(),
                CreatedBy = "system",
                AssignedTo = req.AssignedTo,
                CreatedAt = now,
                UpdatedAt = now,
                DueDate = req.DueDate,
            };

            _tasks[task.Id] = task;
            return task;
        }
        finally
        {
            _lock.ExitWriteLock();
        }
    }

    /// <summary>
    /// Retrieves a task by ID. Throws NotFoundException if not found.
    /// </summary>
    public TaskModel GetTask(string id)
    {
        _lock.EnterReadLock();
        try
        {
            if (!_tasks.TryGetValue(id, out var task))
            {
                throw new NotFoundException($"task with ID {id} not found");
            }

            return task;
        }
        finally
        {
            _lock.ExitReadLock();
        }
    }

    /// <summary>
    /// Updates an existing task with partial update semantics.
    /// Only non-null fields are applied. Sets CompletedAt when status transitions to completed.
    /// Mirrors Go's UpdateTask pointer-field semantics.
    /// </summary>
    public TaskModel UpdateTask(string id, UpdateTaskRequest req)
    {
        _lock.EnterWriteLock();
        try
        {
            if (!_tasks.TryGetValue(id, out var task))
            {
                throw new NotFoundException($"task with ID {id} not found");
            }

            // Update fields if provided (null means "not provided")
            if (req.Title != null)
            {
                task.Title = req.Title;
            }

            if (req.Description != null)
            {
                task.Description = req.Description;
            }

            if (req.Status != null)
            {
                if (!req.Status.Value.IsValid())
                {
                    throw new ValidationException($"invalid status: {req.Status.Value.ToSnakeCase()}");
                }

                task.Status = req.Status.Value;

                // Set completed time if status changes to completed
                if (task.Status == Models.TaskStatus.Completed && task.CompletedAt == null)
                {
                    task.CompletedAt = DateTime.UtcNow;
                }
            }

            if (req.Priority != null)
            {
                if (!req.Priority.Value.IsValid())
                {
                    throw new ValidationException($"invalid priority: {req.Priority.Value.ToSnakeCase()}");
                }

                task.Priority = req.Priority.Value;
            }

            if (req.Tags != null)
            {
                task.Tags = req.Tags;
            }

            if (req.AssignedTo != null)
            {
                task.AssignedTo = req.AssignedTo;
            }

            if (req.DueDate != null)
            {
                task.DueDate = req.DueDate;
            }

            task.UpdatedAt = DateTime.UtcNow;
            return task;
        }
        finally
        {
            _lock.ExitWriteLock();
        }
    }

    /// <summary>
    /// Updates only the status of a task. Delegates to UpdateTask.
    /// </summary>
    public TaskModel UpdateTaskStatus(string id, Models.TaskStatus status)
    {
        return UpdateTask(id, new UpdateTaskRequest
        {
            Status = status
        });
    }

    /// <summary>
    /// Deletes a task by ID. Throws NotFoundException if not found.
    /// </summary>
    public void DeleteTask(string id)
    {
        _lock.EnterWriteLock();
        try
        {
            if (!_tasks.ContainsKey(id))
            {
                throw new NotFoundException($"task with ID {id} not found");
            }

            _tasks.Remove(id);
        }
        finally
        {
            _lock.ExitWriteLock();
        }
    }

    /// <summary>
    /// Returns a paginated, filtered, and sorted list of tasks.
    /// Mirrors Go's ListTasks with identical filtering, sorting, and pagination semantics.
    /// </summary>
    public ListTasksResponse ListTasks(ListTasksQuery query)
    {
        _lock.EnterReadLock();
        try
        {
            // Convert dictionary to list for filtering and sorting
            var tasks = new List<TaskModel>(_tasks.Values);

            // Apply filters
            tasks = FilterTasks(tasks, query);

            // Apply sorting
            SortTasks(tasks, query.SortOrder);

            // Apply pagination
            var totalCount = tasks.Count;
            var pageSize = query.PageSize;
            if (pageSize <= 0)
            {
                pageSize = 20;
            }
            if (pageSize > 100)
            {
                pageSize = 100;
            }

            var startIndex = 0;
            if (!string.IsNullOrEmpty(query.PageToken))
            {
                if (int.TryParse(query.PageToken, out var idx))
                {
                    startIndex = idx;
                }
            }

            var endIndex = startIndex + pageSize;
            if (endIndex > totalCount)
            {
                endIndex = totalCount;
            }

            var paginatedTasks = new List<TaskModel>();
            if (startIndex < totalCount)
            {
                paginatedTasks = tasks.GetRange(startIndex, endIndex - startIndex);
            }

            string? nextPageToken = null;
            if (endIndex < totalCount)
            {
                nextPageToken = endIndex.ToString();
            }

            return new ListTasksResponse
            {
                Tasks = paginatedTasks,
                NextPageToken = nextPageToken,
                TotalCount = totalCount,
            };
        }
        finally
        {
            _lock.ExitReadLock();
        }
    }

    /// <summary>
    /// Applies status, assigned_to, and tags filters to the task list.
    /// Mirrors Go's filterTasks exactly.
    /// </summary>
    private static List<TaskModel> FilterTasks(List<TaskModel> tasks, ListTasksQuery query)
    {
        var filtered = new List<TaskModel>();

        foreach (var task in tasks)
        {
            // Filter by status
            if (!string.IsNullOrEmpty(query.Status))
            {
                if (TaskStatusExtensions.TryParseSnakeCase(query.Status, out var statusFilter))
                {
                    if (task.Status != statusFilter)
                    {
                        continue;
                    }
                }
                else
                {
                    // If the status string doesn't parse, no task will match
                    continue;
                }
            }

            // Filter by assigned user
            if (!string.IsNullOrEmpty(query.AssignedTo))
            {
                if (task.AssignedTo == null || task.AssignedTo != query.AssignedTo)
                {
                    continue;
                }
            }

            // Filter by tags (AND logic - task must have ALL specified tags)
            if (!string.IsNullOrEmpty(query.Tags))
            {
                var tags = query.Tags.Split(',');
                var hasAllTags = true;
                foreach (var tag in tags)
                {
                    var trimmedTag = tag.Trim();
                    if (!task.Tags.Contains(trimmedTag))
                    {
                        hasAllTags = false;
                        break;
                    }
                }
                if (!hasAllTags)
                {
                    continue;
                }
            }

            filtered.Add(task);
        }

        return filtered;
    }

    /// <summary>
    /// Sorts tasks based on the specified sort order.
    /// Uses List.Sort (unstable) to match Go's sort.Slice behavior.
    /// Due date null handling: nulls sort last for both ascending and descending.
    /// </summary>
    private static void SortTasks(List<TaskModel> tasks, string? sortOrder)
    {
        if (string.IsNullOrEmpty(sortOrder))
        {
            return;
        }

        switch (sortOrder)
        {
            case "created_at_asc":
                tasks.Sort((a, b) => a.CreatedAt.CompareTo(b.CreatedAt));
                break;

            case "created_at_desc":
                tasks.Sort((a, b) => b.CreatedAt.CompareTo(a.CreatedAt));
                break;

            case "due_date_asc":
                tasks.Sort((a, b) =>
                {
                    if (a.DueDate == null && b.DueDate == null) return 0;
                    if (a.DueDate == null) return 1;  // null sorts last
                    if (b.DueDate == null) return -1; // null sorts last
                    return a.DueDate.Value.CompareTo(b.DueDate.Value);
                });
                break;

            case "due_date_desc":
                tasks.Sort((a, b) =>
                {
                    if (a.DueDate == null && b.DueDate == null) return 0;
                    if (a.DueDate == null) return 1;  // null sorts last
                    if (b.DueDate == null) return -1; // null sorts last
                    return b.DueDate.Value.CompareTo(a.DueDate.Value);
                });
                break;

            case "priority_asc":
                tasks.Sort((a, b) => a.Priority.GetValue().CompareTo(b.Priority.GetValue()));
                break;

            case "priority_desc":
                tasks.Sort((a, b) => b.Priority.GetValue().CompareTo(a.Priority.GetValue()));
                break;

            // Default: no sorting (matches Go behavior when sort_order is empty or unrecognized)
        }
    }

    /// <summary>
    /// Creates two sample tasks on startup, matching Go's initializeSampleData exactly.
    /// </summary>
    private void InitializeSampleData()
    {
        var sampleTasks = new List<CreateTaskRequest>
        {
            new()
            {
                Title = "Implement Go REST API",
                Description = "Create REST API with Gin framework",
                Priority = TaskPriority.High,
                Tags = new List<string> { "go", "rest", "api" },
            },
            new()
            {
                Title = "Add gRPC support",
                Description = "Implement gRPC server with native Go support",
                Priority = TaskPriority.Medium,
                Tags = new List<string> { "go", "grpc", "protobuf" },
            },
        };

        foreach (var req in sampleTasks)
        {
            CreateTask(req);
        }
    }
}
